File: database/rdbms/algorithm/join/complexity/python/join_simulator/data/table.py
"""テーブルとインデックスのデータ構造

TableRecord, TableData, IndexDataを提供
"""

from dataclasses import dataclass
from typing import Dict, List, Optional, Iterator
import bisect


@dataclass
class TableRecord:
    """テーブルレコード

    Attributes:
        primary_key: 主キー
        filter_key: WHERE条件で使用するフィルタキー
        column_value: JOIN条件で使用するカラム値
    """
    primary_key: int
    filter_key: int
    column_value: int


class TableData:
    """テーブルデータ

    レコードのリストと主キーインデックスを保持。
    主キーによる高速な検索をサポート。
    """

    def __init__(self, capacity: int = 0) -> None:
        """初期化

        Args:
            capacity: 初期容量のヒント（最適化用）
        """
        self._records: List[TableRecord] = []
        self._pk_index: Dict[int, int] = {}  # primary_key -> list index

    def insert(self, record: TableRecord) -> None:
        """レコードを挿入

        Args:
            record: 挿入するレコード
        """
        idx = len(self._records)
        self._pk_index[record.primary_key] = idx
        self._records.append(record)

    def get_by_pk(self, pk: int) -> Optional[TableRecord]:
        """主キーでレコードを取得

        Args:
            pk: 主キー

        Returns:
            レコード、見つからない場合はNone
        """
        idx = self._pk_index.get(pk)
        return self._records[idx] if idx is not None else None

    def iter(self) -> Iterator[TableRecord]:
        """全レコードをイテレート

        フルスキャン用のイテレータを返す

        Returns:
            レコードのイテレータ
        """
        return iter(self._records)

    def __len__(self) -> int:
        """レコード数を取得

        Returns:
            レコード数
        """
        return len(self._records)


class IndexData:
    """Bツリーインデックスの簡易シミュレーション

    bisectモジュールを使用してBツリーのO(log n)検索特性をシミュレート。
    ソート済みキーリストと、キーから主キーリストへのマッピングを保持。
    """

    def __init__(self) -> None:
        """初期化"""
        # ソート済みキーリスト（Bツリーのソート特性をシミュレート）
        self._keys: List[int] = []
        # キー -> 主キーリストのマッピング
        self._index: Dict[int, List[int]] = {}

    def insert(self, filter_key: int, primary_key: int) -> None:
        """インデックスエントリを追加

        Args:
            filter_key: インデックスキー
            primary_key: 対応する主キー
        """
        if filter_key not in self._index:
            self._index[filter_key] = []
            # bisect.insort()でソート済み状態を維持（Bツリーのソート特性）
            bisect.insort(self._keys, filter_key)
        self._index[filter_key].append(primary_key)

    def search(self, filter_key: int) -> Optional[List[int]]:
        """インデックス検索（Bツリー探索をシミュレート）

        bisect.bisect_left()を使用してO(log n)の二分探索を実行。
        これによりBツリーの探索特性をシミュレート。

        Args:
            filter_key: 検索キー

        Returns:
            一致する主キーのリスト、見つからない場合はNone
        """
        # bisect.bisect_left()でO(log n)二分探索を実行
        pos = bisect.bisect_left(self._keys, filter_key)
        if pos < len(self._keys) and self._keys[pos] == filter_key:
            return self._index[filter_key]
        return None

    def __len__(self) -> int:
        """インデックスエントリの総数を取得

        Returns:
            全キーに対する主キーの総数
        """
        return sum(len(pks) for pks in self._index.values())

File: database/rdbms/algorithm/join/complexity/rust/src/data/table.rs
use std::collections::BTreeMap;

#[derive(Debug, Clone)]
pub struct TableRecord {
    pub primary_key: u64,
    pub filter_key: u64,
    pub column_value: u64,
}

#[derive(Debug)]
pub struct TableData {
    records: Vec<TableRecord>,
    pk_index: BTreeMap<u64, usize>,
}

impl TableData {
    pub fn new(capacity: usize) -> Self {
        Self {
            records: Vec::with_capacity(capacity),
            pk_index: BTreeMap::new(),
        }
    }

    pub fn insert(&mut self, record: TableRecord) {
        let idx = self.records.len();
        self.pk_index.insert(record.primary_key, idx);
        self.records.push(record);
    }

    pub fn get_by_pk(&self, pk: u64) -> Option<&TableRecord> {
        self.pk_index.get(&pk).and_then(|&idx| self.records.get(idx))
    }

    pub fn iter(&self) -> impl Iterator<Item = &TableRecord> {
        self.records.iter()
    }

    pub fn len(&self) -> usize {
        self.records.len()
    }

    #[allow(dead_code)]
    pub fn is_empty(&self) -> bool {
        self.records.is_empty()
    }
}

#[derive(Debug)]
pub struct IndexData {
    index: BTreeMap<u64, Vec<u64>>,
}

impl IndexData {
    pub fn new() -> Self {
        Self {
            index: BTreeMap::new(),
        }
    }

    pub fn insert(&mut self, filter_key: u64, primary_key: u64) {
        self.index
            .entry(filter_key)
            .or_insert_with(Vec::new)
            .push(primary_key);
    }

    pub fn search(&self, filter_key: u64) -> Option<&Vec<u64>> {
        self.index.get(&filter_key)
    }

    #[allow(dead_code)]
    pub fn len(&self) -> usize {
        self.index.values().map(|v| v.len()).sum()
    }

    #[allow(dead_code)]
    pub fn is_empty(&self) -> bool {
        self.index.is_empty()
    }
}

impl Default for IndexData {
    fn default() -> Self {
        Self::new()
    }
}
